autoload/neomake.vim

```diff
@@ -117,39 +123,75 @@ function! s:MakeJob(make_id, maker) abort
 
     try
         let has_args = type(args) == type([])
-        if has('nvim')
-            let argv = [exe]
-            if has_args
-                let argv = argv + args
-            endif
-            call neomake#utils#LoudMessage('Starting: '.join(argv, ' '))
-            let opts = {
-                \ 'on_stdout': function('neomake#MakeHandler'),
-                \ 'on_stderr': function('neomake#MakeHandler'),
-                \ 'on_exit': function('neomake#MakeHandler')
-                \ }
-            let job = jobstart(argv, opts)
-            let jobinfo.start = localtime()
-            let jobinfo.last_register = 0
-
-            if job == 0
-                throw 'Job table is full or invalid arguments given'
-            elseif job == -1
-                throw 'Non executable given'
-            endif
+        if neomake#has_async_support()
+            if has('nvim')
+                let argv = [exe]
+                if has_args
+                    let argv = argv + args
+                endif
+                call neomake#utils#LoudMessage('Starting: '.join(argv, ' '))
+                let opts = {
+                            \ 'on_stdout': function('neomake#MakeHandler'),
+                            \ 'on_stderr': function('neomake#MakeHandler'),
+                            \ 'on_exit': function('neomake#MakeHandler')
+                            \ }
+                let job = jobstart(argv, opts)
+                let jobinfo.start = localtime()
+                let jobinfo.last_register = 0
+
+                if job == 0
+                    throw 'Job table is full or invalid arguments given'
+                elseif job == -1
+                    throw 'Non executable given'
+                endif
 
-            let jobinfo.id = job
-            let s:jobs[job] = jobinfo
-            let maker_key = s:GetMakerKey(a:maker)
-            let s:jobs_by_maker[maker_key] = jobinfo
-            call s:AddJobinfoForCurrentWin(jobinfo.id)
-            let r = jobinfo.id
-            if !exists('s:jobids_by_makeid[a:make_id]')
-                let s:jobids_by_makeid[a:make_id] = []
+                let jobinfo.id = job
+                let s:jobs[job] = jobinfo
+                let maker_key = s:GetMakerKey(a:maker)
+                let s:jobs_by_maker[maker_key] = jobinfo
+                call s:AddJobinfoForCurrentWin(jobinfo.id)
+                let r = jobinfo.id
+                if !exists('s:jobids_by_makeid[a:make_id]')
+                    let s:jobids_by_makeid[a:make_id] = []
+                endif
+                call add(s:jobids_by_makeid[a:make_id], jobinfo.id)
+            else
+                if has_args
+                    if neomake#utils#IsRunningWindows()
+                        let program = [exe.' '.join(map(args, 'v:val'))]
+                    else
+                        let program = ['/bin/sh', '-c', exe.' '.join(map(args, 'shellescape(v:val)'))]
```

Comment:
I've been testing it without this wrapping shell, but without this "trick" it seems that it doesn't work... I did some research but I couldn't find any hint about the reasoning behind this. Projects like vim-gutter and agrep are taking the same approach. Could it be a Vim's bug? In the meantime, do you think a simple comment like this description would be good enough?